### ISSUE TITLE
How long does it takes to reproduce the results in the paper?

### ISSUE THREAD

**Issue by Neko9810** [changdaeoh/BlackVIP#1]
We are interested in your work, and we want to reproduce the results. But we find that it runs extremely slow, e.g., about 2 days 8h with a single RTX3090.

We set CFG=vit_b16 and ptb=vit-base, and others followed the configuration.md.

Is something wrong?

**Comment by changdaeoh** [comment]
Thank you for showing interest in our research!

Unfortunately, black-box optimization algorithms (including ours or baseline) require huge iterations for convergence than white-box true gradient-based optimizers. 
Moreover, for stable gradient approximation, they practically require K-time-repeated gradient estimates for each iteration, and we use average of them. (K = 5 by default) 

One practical solution is replacing the function `self.spsa_grad_estimate_bi( )` of `trainers/blackvip.py `as below: (a variant modifying the approximation type from two-side to one-side for faster experiments)

```python
#* repeat k times and average them for stabilizing
        output = self.model(image)
        loss_pivot = self.loss_fn(output, label)
        ghats, ighats, igrghat = [], [], 0
        for spk in range(self.sp_avg):
            #! Bernoulli {-1, 1}
            # perturb = torch.bernoulli(torch.empty(self.N_params).uniform_(0,1)).cuda()
            # perturb[perturb < 1] = -1
            #! Segmented Uniform [-1, 0.5] U [0.5, 1]
            p_side = (torch.rand(self.N_params).reshape(-1,1) + 1)/2
            samples = torch.cat([p_side,-p_side], dim=1)
            perturb = torch.gather(samples, 1, torch.bernoulli(torch.ones_like(p_side)/2).type(torch.int64)).reshape(-1).cuda()
            del samples; del p_side

            #* one-side Approximated Numerical Gradient
            w_r = w + ck*perturb
            torch.nn.utils.vector_to_parameters(w_r, self.model.coordinator.dec.parameters())
            output_pt = self.model(image)
            loss_pt = self.loss_fn(output_pt, label)
            #* parameter update via approximated gradient
            ghat = (loss_pt - loss_pivot)/(ck*perturb)
            ghats.append(ghat.reshape(1, -1))

        if self.sp_avg == 1: pass
        else: ghat = torch.cat(ghats, dim=0).mean(dim=0)
        acc = (compute_accuracy(output, label)[0]).item()
        return ghat, loss_pivot, acc, igrghat
```
But, this might decrease the gradient approximation quality resulting in lower performance. For faster validation, you can consider other smaller dataset, such as EuroSAT or SVHN, that has smaller class-label categories.
